src/fp-ts/createSetFromArray.ts

```diff
@@ -0,0 +1,38 @@
+import * as t from 'io-ts'
+import { insert, toArray } from 'fp-ts/lib/Set'
+import { array } from 'fp-ts/lib/Array'
+import { Ord } from 'fp-ts/lib/Ord'
+
+export class SetFromArrayType<RT extends t.Any, A = any, O = A, I = t.mixed> extends t.Type<A, O, I> {
+  readonly _tag: 'SetFromArrayType' = 'SetFromArrayType'
+  constructor(
+    name: string,
+    is: SetFromArrayType<RT, A, O, I>['is'],
+    validate: SetFromArrayType<RT, A, O, I>['validate'],
+    serialize: SetFromArrayType<RT, A, O, I>['encode'],
+    readonly type: RT
+  ) {
+    super(name, is, validate, serialize)
+  }
+}
+
+export const createSetFromArray = <RT extends t.Type<A, O>, A = any, O = any>(
```

Comment:
- `is` should check the elements other that the constructor (with `Set.every` I guess). Also would be better to add a related test
- add `readonly ordA: Ord<t.TypeOf<RT>>` to `SetFromArrayType`?
- in `createSetFromArray` `RT` should extends `t.Mixed` and return a `SetFromArrayType<RT, Set<t.TypeOf<RT>>, Array<t.OutputOf<RT>>, t.mixed>` instead of a generic `Type`
- `insertSetOrdA` looks slow since return a new Set every time, we could just use mutability here

since code worths thousands words:

```ts
export class SetFromArrayType<RT extends t.Any, A = any, O = A, I = t.mixed> extends t.Type<A, O, I> {
  readonly _tag: 'SetFromArrayType' = 'SetFromArrayType'
  constructor(
    name: string,
    is: SetFromArrayType<RT, A, O, I>['is'],
    validate: SetFromArrayType<RT, A, O, I>['validate'],
    serialize: SetFromArrayType<RT, A, O, I>['encode'],
    readonly type: RT,
    readonly ordA: Ord<t.TypeOf<RT>>
  ) {
    super(name, is, validate, serialize)
  }
}

export const createSetFromArray = <RT extends t.Mixed>(
  type: RT,
  ordA: Ord<t.TypeOf<RT>>,
  name: string = `Set<${type.name}>`
): SetFromArrayType<RT, Set<t.TypeOf<RT>>, Array<t.OutputOf<RT>>, t.mixed> => {
  const ArrayType = t.array(type)
  const equals = ordA.equals
  return new SetFromArrayType(
    name,
    (m): m is Set<t.TypeOf<RT>> => m instanceof Set && every(m, type.is),
    (m, c) => {
      const validation = ArrayType.validate(m, c)
      if (validation.isLeft()) {
        return validation as any
      } else {
        const as = validation.value
        const len = as.length
        const r = new Set()
        for (let i = 0; i < len; i++) {
          const a = as[i]
          if (!some(r, x => equals(x, a))) {
            r.add(a)
          }
        }
        return r.size !== len ? t.failure(as, c) : t.success(r)
      }
    },
    toArray(ordA),
    type,
    ordA
  )
}
```

Reply:
@gcanti from what we've discovered, using `<RT extends t.Mixed>` is inherently unsafe;
if you replace `<RT extends t.Mixed>` by `<RT extends t.Type<A, O>, A = any, O = any>` in your example; you immediately see it does not typecheck.

Indeed, it errors on 'encode' (with reasons); the correct code for encode should be:
`s => ArrayType.encode(toArray(ordA)(s)),`

So I'm really challenging how to write the type params in the lib as having unchecked code should be a deal breaker IMHO.

Let me know your thinking!

Here's the updated proposal:

```typescript

export class SetFromArrayType<RT extends t.Any, A = any, O = A, I = t.mixed> extends t.Type<A, O, I> {
  readonly _tag: 'SetFromArrayType' = 'SetFromArrayType'
  constructor(
    name: string,
    is: SetFromArrayType<RT, A, O, I>['is'],
    validate: SetFromArrayType<RT, A, O, I>['validate'],
    serialize: SetFromArrayType<RT, A, O, I>['encode'],
    readonly type: RT,
    readonly ordA: Ord<t.TypeOf<RT>>
  ) {
    super(name, is, validate, serialize)
  }
}

export const createSetFromArray = <RT extends t.Type<A, O>, A = any, O = A>(
  type: RT,
  ordA: Ord<t.TypeOf<RT>>,
  name: string = `Set<${type.name}>`
): SetFromArrayType<RT, Set<t.TypeOf<RT>>, Array<t.OutputOf<RT>>, t.mixed> => {
  const ArrayType = t.array(type)
  const equals = ordA.equals
  return new SetFromArrayType(
    name,
    (m): m is Set<t.TypeOf<RT>> => m instanceof Set && every(m, type.is),
    (m, c) => {
      const validation = ArrayType.validate(m, c)
      if (validation.isLeft()) {
        return validation as any
      } else {
        const as = validation.value
        const len = as.length
        const r = new Set<t.OutputOf<RT>>()
        for (let i = 0; i < len; i++) {
          const a = as[i]
          if (!some(r, x => equals(x, a))) {
            r.add(a)
          }
        }
        return r.size !== len ? t.failure(as, c) : t.success(r)
      }
    },
    s => ArrayType.encode(toArray(ordA)(s)),
    type,
    ordA
  )
}
```

Reply:
Also @gcanti 

`assert.deepEqual(T.is(new Set([1, 1])), false) // Should we?` (perf killer)

Reply:
> if you replace <RT extends t.Mixed> by <RT extends t.Type<A, O>, A = any, O = any> in your example; you immediately see it does not typecheck

Oh no I forgot about that.. my bad, thanks for pointing out (then should we change `createOptionFromNullable` accordingly?).

```
assert.deepEqual(T.is(new Set([1, 1])), false) // Should we?
```

No, why should be `false`?

Reply:
@gcanti should not be false as new Set([1, 1]) === new Set([1]) (I've added the test, that's ok)

Reply:
@gcanti yes `createOptionFromNullable` should be changed that way, there may be other places where there's hidden inconsistencies `*-ts`?

Reply:
@sledorze Not sure, IIRC the `<RT extends t.Mixed>` pattern is only used in `io-ts` but is well tested so should not be a problem